Cargill/splinter
# Add permissions for REST API authorization

This PR adds permission strings to all Splinter REST API endpoints and updates the REST API's authorization guard to start using these permission strings for authorization checks.

This PR should not change any functionality for the end user, but it will enable restricting endpoints based on permissions in a future PR.

#### Testing

1. Start splinterd with Biome enabled

```bash
$ mkdir /tmp/splinter
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- database migrate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path cli/Cargo.toml -- cert generate
$ SPLINTER_HOME=/tmp/splinter cargo run --manifest-path splinterd/Cargo.toml --features experimental -- --enable-biome --tls-insecure -vv
```

2. In a new terminal window, register and login a user to verify the "allow all" endpoints work properly:

```bash
$ curl -X POST -H 'Content-Type: application/json' -d '{"username": "user", "hashed_password": "123"}' http://localhost:8080/biome/register
$ curl -X POST -H 'Content-Type: application/json' -d '{"username": "user", "hashed_password": "123"}' http://localhost:8080/biome/login
```

3. Verify that other endpoints still require an authorization header by attempting to access Biome keys without authorization:

```bash
$ curl http://localhost:8080/biome/keys
```

4. Use the token from your login in step 2 to verify you can get Biome keys with correct authorization:

```bash
$ curl -H 'Authorization: Bearer Biome:<token>' http://localhost:8080/biome/keys
```
